dodiscover/constraint/pcalg.py

```diff
@@ -0,0 +1,496 @@
+import logging
+from itertools import combinations, permutations
+from typing import Any, Dict, List, Optional, Set
+
+import networkx as nx
+from pywhy_graphs import CPDAG
+
+from dodiscover import Context
+from dodiscover.ci.base import BaseConditionalIndependenceTest
+from dodiscover.constraint.utils import is_in_sep_set
+
+from ._classes import BaseConstraintDiscovery
+from ._protocol import EquivalenceClassProtocol
+
+logger = logging.getLogger()
+
+
+class PC(BaseConstraintDiscovery):
```

Comment:
Wouldn't it make more sense to follow-up on the effort to bring the '[causal-learn' package](https://github.com/cmu-phil/causal-learn) into PyWhy instead of re-implementing the functionalities?

Reply:
These were refactored from my package, `causal-networkx`. From my exp. the `causal-learn` package was very hard to refactor into a working extensible version. Moreover, the package lacked [unit-tests](https://github.com/cmu-phil/causal-learn/blob/main/tests/TestFCI.py), so I would be wary of copy-pasting here directly w/o extensive testing. 

Is there any reason this approach is not desired?

**Just as background:** Not trying to be too opinionated btw. The package is great for doing some initial applications on the problem I was working on, but I was also working on extending structure learning and researching modifications and it was very difficult for me to do so. Hence I re-implemented the abstractions from scratch.

Reply:
I was more thinking along the line that we plan to integrate the causal-learn package in the future anyway (at least, we are talking to the authors of the package) and then it might be better to rather try to refactor their package and make it compatible (rather then re-implementing methods that already exist in many other packages).

Also keep in mind that we can not simply copy over code and refactor it, that would violate the license :)

Reply:
I see! Makes sense if they are on board to try. I also took inspiration from their code as well. This implementation was to address the action item we discussed last time in the meeting of getting an end-to-end PC example using the graphs to be PRed into networkx. 

It was difficult for me to parse https://github.com/cmu-phil/causal-learn/blob/main/causallearn/search/ConstraintBased/PC.py to see if we're missing anything in the core pc alg, so lmk. I think the PR's implementation follows the abstraction of the PC algorithm, so it should be very easy to add functionality at any stage of the algorithm if we are missing something: e.g. the abstraction has the following steps for all constraint-based algorithms:

1. initialization
2. learning a skeleton (this is dedicated to a separate algorithm altogether, since there are many proposals in the literature for improving this process
3. orienting unshielded colliders 
4. orienting edges

(for example, here it is demonstrated that one can easily implement Conservative/MaxVotePC by overriding one method: the orientation of unshielded colliders -> 80 LOC). The algorithm then works as expected locally (unit tests to be added).

https://github.com/cmu-phil/causal-learn/blob/main/causallearn/search/ConstraintBased/PC.py does show that there is an algorithm missing, which is the `mvpc` algo for missing-value. I think that involves modification of the skeleton procedure IIUC and then everything else should work as expected.

How does that sound?

Reply:
Sounds good. Its also nothing crucial, I just want to make sure that you don't have double work if we can leverage other packages :) I think the PC algorithm is a good example to see how the required APIs could look like.

Reply:
I agree that I'm happy to get this in there now.  The causal-learn team did indicate they would change the license but we shouldn't block on them.  

I would like to see this PR broken up though (see message below) so we can commit these parts piecemeal?